### PR TITLE
yarn react build|dev

### PR DESCRIPTION
**Pull request by ericclemmons**
- [x] Replace #177 
- [x] Replace #186 
- [x] Use `tsup` for bundling – https://tsup.egoist.sh/
- [x] Output `@aws-amplify/ui-react/styles.css`
- [x] Inline all `@aws-amplify/*` dependencies (so it's a single bundle)
- [x] Remove Next.s compilation for `packages/*`
- [x] `yarn react dev` for development
- [x] `yarn react build` for production build
- [x] `yarn core build`
- [x] Watch changes in `@aws-amplify/ui-core` when doing `yarn react dev`

```
❯ yarn react dev
yarn run v1.22.10
$ yarn workspace @aws-amplify/ui-react dev
$ rimraf dist
$ tsup --watch src --watch ../core/src
 CLI  Building entry: src/index.tsx
 CLI  Using tsconfig: /Users/ecclemm/Projects/aws-amplify/amplify-ui/packages/react/tsconfig.json
 CLI  tsup v4.14.0
 CLI  Using tsup config: /Users/ecclemm/Projects/aws-amplify/amplify-ui/packages/react/tsup.config.cjs
 CLI  Running in watch mode
 CLI  Target: node12
 CJS  Build start
 ESM  Build start
 ESM  Build success in 6788ms
 CJS  Build success in 7019ms
 CLI  Watching for changes in "src" | "../core/src"
 CLI  Ignoring changes in "**/{.git,node_modules}/**" | "dist"
 DTS  Build start
 DTS  Build success in 28272ms
 CLI  Change detected: change ../core/src/authService.ts
 CJS  Build start
 ESM  Build start
 CJS  Build success in 1408ms
 ESM  Build success in 1407ms
```

### CODE REVIEW

File: packages/react/package.json
Comment (ericclemmons) [inline packages/react/package.json:10]: `import "@aws-amplify/ui-react/styles.css"` correctly resolves to `dist/index.css`.

Neat, right?

File: packages/react/package.json
Comment (ericclemmons) [inline packages/react/package.json:50]: Build tools like `microbundle`, `tsup`, etc. externalize `dependencies` and inline `devDependencies`.

This is helpful for using packages that we don't want to publish.

File: packages/react/patches/tsup+4.14.0.patch
Comment (ericclemmons) [inline packages/react/patches/tsup+4.14.0.patch:9]: https://esbuild.github.io/content-types/#auto-import-for-jsx

File: packages/react/src/index.tsx
Comment (ericclemmons) [inline packages/react/src/index.tsx:1]: This is really nice – building (via PostCSS) styles as part of a single package & process.

File: packages/theme-base/package.json
Comment (ericclemmons) [inline packages/theme-base/package.json:27]: The bundler had issues with WOFF files inside of this dependency.

Admittedly, this is weird to me as part of a library.  Because fonts require asset files (e.g. `.woff`, `.ttf`, etc.), these have to be bundled by the client app.

Normally, I see styles only specify `font-family` and leave it to the consumer to determine **where** the fonts are located:

- https://material-ui.com/customization/typography/#self-hosted-fonts
- https://tailwindui.com/documentation#optional-add-the-inter-font-family
- https://create-react-app.dev/docs/adding-images-fonts-and-files/
- https://github.com/csstools/postcss-font-magician

Because fonts can be very expensive, I think we should revisit #165 and try to make bundle-size as much of an opt-in as possible, like https://tailwindui.com/documentation#optional-add-the-inter-font-family.

If a customer doesn't want it, `font-family` will naturally go to the next available font.

If a customer wants to load it in a blocking way, they can use `<link>`.

If a customer wants to load it in a non-blocking way, they can do what Next.js does:
> https://nextjs.org/docs/basic-features/font-optimization

File: packages/react/package.json
Comment (dbanksdesign) [reply packages/react/package.json:10]: neat

File: packages/react/package.json
Comment (dbanksdesign) [reply packages/react/package.json:50]: Broader question: do we want to publish ui-theme-base? I noticed we publish @aws-amplify/ui which is kind of similar to what ui-theme-base is now. I don't have a strong opinion, just wanted to bring it up. Should we rename ui-theme-base to just ui?

File: package.json
Comment (ericclemmons) [inline package.json:4]: Going to replace this with `turbo`

Comment (dbanksdesign) [inline package.json:4]: ohh I like, now I can just run install and then start up the docs or example next server

File: package.json
Comment (ericclemmons) [inline package.json:41]: `tsup` checked all the boxes that `vite` did, + `.d.ts` output.

`Nx` requires a ton of convention, which is **great**, but more than I wanted to change.

`microbundle` is slow because of rollup.

File: packages/core/package.json
Comment (ericclemmons) [inline packages/core/package.json:28]: Otherwise our `amazon-cognito-identity-js` and `aws-amplify` dependencies get bundled!

File: packages/react/package.json
Comment (ericclemmons) [inline packages/react/package.json:32]: I'll move theme-base into core in another PR. We won't lose tree-shaking or anything, but it'll be 1 less bundle for customers to keep in sync.

File: packages/react/package.json
Comment (ericclemmons) [inline packages/react/package.json:42]: This way we don't bundle dependencies.

File: packages/react/src/index.tsx
Comment (ericclemmons) [reply packages/react/src/index.tsx:1]: Somehow the SCSS stuff just worked?? I need to check the docs to be sure...

File: packages/react/src/globals.js
Comment (ericclemmons) [inline packages/react/src/globals.js:10]: Vite does this automatically. I had to patch `tsup` to avoid changing EVERY `.tsx` file.

File: packages/react/tsconfig.json
Comment (ericclemmons) [inline packages/react/tsconfig.json:4]: I thought this was the default, but `tsup` & `vite` weren't picking up on `node_modules` until I added this.

**Review by dbanksdesign — APPROVED** [review]
Overall, I love it. A couple small questions, nothing blocking.

File: packages/react/tsconfig.json
Comment (ericclemmons) [inline packages/react/tsconfig.json:12]: So, TIL : https://stackoverflow.com/questions/41993811/understanding-target-and-module-in-tsconfig

`module` basically means `module.exports` or `import/export` or `import()`:
> https://github.com/microsoft/TypeScript/issues/24082

File: packages/react/tsconfig.json
Comment (ericclemmons) [inline packages/react/tsconfig.json:13]: `target` means the language features:
> https://stackoverflow.com/questions/41993811/understanding-target-and-module-in-tsconfig

Targeting evergreen browsers, this could even be `esnext` but I figure we'll standup more examples (CRA, Nuxt, etc.) to determine what that impact is.

But the less TS has to transpile, the faster our builds will be.

File: packages/react/tsup.config.cjs
Comment (ericclemmons) [inline packages/react/tsup.config.cjs:4]: I did it this way for type safety vs. `tsup src/index.tsx --format cjs,esm ...`

File: patches/tsup+4.14.0.patch
Comment (ericclemmons) [inline patches/tsup+4.14.0.patch:9]: https://github.com/egoist/tsup/issues/390

File: packages/core/package.json
Comment (wlee221) [inline packages/core/package.json:13]: Should we have a separate `yarn clean` script here? I find explicit clean scripts when I'm looking at module/build issues.

```suggestion
    "clean": "rimraf dist",
    "prebuild": "yarn clean",
```

File: patches/tsup+4.14.0.patch
Comment (wlee221) [inline patches/tsup+4.14.0.patch:1]: TIL https://www.npmjs.com/package/patch-package

**Review by wlee221 — APPROVED** [review]
This is looking great! Thanks @ericclemmons. Just got some suggestions/questions.

File: CONTRIBUTING.md
Comment (ericclemmons) [reply CONTRIBUTING.md:24]: `yarn core dev` is the answer for local development (including watching).

(I also learned that the build tool may not always be the dev tool!)

File: package.json
Comment (ericclemmons) [reply package.json:4]: Exactly right @dbanksdesign – I want the majority of the packages/examples/docs/etc. to "just work" with `git clone` and `yarn install`.

The only exception is really `environments/`, and that's because they require an AWS account.

In the future, we can even avoid that by including `aws-exports.js` – we would only need to monitor those resources for usage/abuse.

File: packages/core/package.json
Comment (ericclemmons) [reply packages/core/package.json:13]: You know, I went back-and-forth on this.  What I noticed was that it was tough to trace when something was being used and where. (`build:icons` is an example of this)

Having `clean` explicitly defined kinda felt like I was promoting it to a public API (e.g. `yarn react clean`) vs. having only `build|dev` being the public APIs.

I hope the indirection from `build => prebuild => clean` is clearer with `build => prebuild` 🙏 

(I don't have a strong opinion – I've done NPM scripts every which way since [per-env](https://github.com/ericclemmons/per-env))

File: packages/react/src/index.tsx
Comment (ericclemmons) [reply packages/react/src/index.tsx:1]: Good question @dbanksdesign!  No, `import '@aws-amplify/ui-react/styles.css'` is still required.

What `import` does here is allow the bundler (`tsup`, `microbundle`, whatever) automatically output both JS & CSS with a single process. (Compared to having to watch CSS **and** JS separately)

This is consistent with how, say, an application is bundled with webpack: even though your app has JS, CSS, SVGs, PNGs, whatever, the bundler identifies all dependencies and outputs them based on your configuration.

**Review by ericclemmons — COMMENTED** [review]
Responding to comments, but somehow it's a review? ¯\_(ツ)_/¯

File: packages/core/package.json
Comment (dbanksdesign) [reply packages/core/package.json:13]: Is there a use-case to run _only_ a clean? I've realized whenever I'm doing a clean it's because I want to blow up the build artifacts and rebuild them, so just being part of `prebuild` makes sense to me. But if there are use cases for "clean and not build" I'm for it.

File: packages/react/package.json
Comment (ericclemmons) [reply packages/react/package.json]: No, any cross-platform/framework dependencies (like the state machine, design tokens, or even specs (we ran into an issue with that!)) ideally lives in `@aws-amplify/ui-core`.

But you're right to ask about [`@aws-amplify/ui`](https://github.com/aws-amplify/amplify-js/tree/main/packages/amplify-ui).

`ui-core` would be a new package, but I don't think it makes sense to have.  `ui-core` is kinda our equivalent of `ui`, so you're right – maybe we rename & merge `ui-core` and `theme-base` into `ui`.

The only thing that I found weird is `aws-amplify` installs it by default:
https://github.com/aws-amplify/amplify-js/blob/da8ef238928fdcad7ab307350ffff8448eec5b26/packages/aws-amplify/package.json#L47

But I have no idea why, or see the need for it.

The more I type it out, the more it makes sense for `@aws-amplify/ui` to contain our core/shared assets and `@aws-amplify/ui-*` be framework-specific stuff.  `core` just seemed to be common convention in the ecosystem (e.g. `@babel/core`).

Let's queue this up for another cleanup task.

File: packages/core/package.json
Comment (ericclemmons) [reply packages/core/package.json:13]: Great question. I don't have one.  Your use-case models my experience as well:

- "Build for production", which means a clean slate so we don't re-publish old artifacts.
- "Dev" _should not_ remove clean because I found out it breaks hot-reload when those references disappear.

I suppose there could be a reason to clean and do dev, but IME `yarn install` is basically our starting point for "setup and build everything, and I'll chose what part I'm working on".

**Pull request merged**
